rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,401 @@
+although there are some odd instances with >100k users. So we consider two options at the moment (see the [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#)):
+- Store notification state separately from a notification. State for new notifications created for every user when they change the state.
+```typescript
+interface  NotificationState {
+  notification_id: string;
+  recipient_id: string;
+  is_read: boolean;
+  is_pinned: boolean;
+}
+``` 
+This option consumes less storage space but might lead to expensive read operations due to an additional JOIN call to search for notification state objects. We don’t expect Kibana to read more than 10-20 notifications at once, so the overhead might be acceptable. Also, this option doesn't rely on built-in ES functionality, which can simplify migration to SavedObject storage later.
+- Store user-specific `NotificationState` as [nested](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-nested-query.html) field of `Notification`. This option can affect ES performance since it reindexes the whole document on every `nested` object update. It can lead to conflicts when several users update the notification state at the same time (see [Best suited for data that does not change frequently](https://www.elastic.co/blog/managing-relations-inside-elasticsearch))
+- Store user-specific `NotificationState` as [child](https://www.elastic.co/guide/en/elasticsearch/reference/current/parent-join.html) of `Notification` object. This option doesn't have the reindexing problem typical of the previous option. But [you cannot sort the results of a has_child/has_parent query using standard sort options.](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-parent-query.html).
+- Create a copy for every notification in a user-specific list of notifications. Denormalization is [the recommended way to manage relationships](https://www.elastic.co/blog/managing-relations-inside-elasticsearch),
+but it leads to duplicating notification messages. 
+
+Stale notifications and their states must be removed from the storage. There should be a dedicated task performing
+look up through the storage to remove notifications with *expire_at* > the current timestamp.
+
+Data are stored in the hidden system index to prevent accidental access by a random user.
+We don’t expect notifications to contain sensitive information.
+They must be stored in an encrypted form otherwise. 
+Kibana can query user-specific notifications by *recipent_id*.
+
+Alternative implementation might be to use SO with OLS([#17888](https://github.com/elastic/kibana/issues/17888)) and built-in migration support. The main obstacle to use SO is limited search functionality that doesn’t allow using complex search queries (can be addressed in [#84729](https://github.com/elastic/kibana/issues/84729)).
```

Comment:
What complex search query would you like to use that saved-objects don't support? We really shouldn't have to be reinventing Kibana-specific storage in situations like these.

Reply:
Search for user notifications and global notifications is alright. I'm concerned about the necessity to use aggregations.
In the Notification Flyout we need to show how many notifications are in a group:
![107675293-3a11b580-6c98-11eb-841c-5b219885d013](https://user-images.githubusercontent.com/3198181/108371594-33d28a80-71fe-11eb-88a9-8401c2be9168.png)


At the later stage we will have Notification Center UI that is full of different counters:
![2021-02-18_13-59-21](https://user-images.githubusercontent.com/3198181/108367040-4f876200-71f9-11eb-80ee-6b888b0331a6.png)

To complicate the calculation logic, notifications have a concept of groups, which should be count as a single notification. (cc @thesmallestduck is it correct?). I was counting on us being able to use bucket aggregations for this case. 
Without aggregations, we have to implement these calculations manually, which can lead to performance problems. 
Do you see any better alternatives? We can start with an implementation based on SO, implementing performance testing from the very beginning to make sure we don't overwhelm Kibana and ES server.

Reply:
Discussed offline with @kobelb: It's okay to start with SO, querying `.kibana` index directly until SO supports advanced search functionality and aggregations (presumably can be done in https://github.com/elastic/kibana/issues/84729)